jumpscale/sals/process/__init__.py

```diff
@@ -84,595 +141,833 @@ def execute(
+        psutil.Process or None: psutil.Process object if he given PID is found, None otherwise.
     """
     try:
         return psutil.Process(pid)
-    except psutil.NoSuchProcess as e:
+    except (psutil.ZombieProcess, psutil.AccessDenied, psutil.NoSuchProcess) as e:
+        # when you query processess owned by another user, especially on macOS and Windows you may get AccessDenied exception
         if die:
             raise e
         else:
             return None
 
 
 def get_user_processes(user):
-    """Get all process for a specific user
+    """Get all process for a specific user.
 
-    Arguments:
-        user {str} -- username
+    Args:
+        user (str): Te user name to match against.
 
-    Returns:
-        [list(int)] -- list of process pids for that user
+    Yields:
+        psutil.Process: psutil.Process object for all processes owned by `user`.
     """
-    result = []
-    for process in psutil.process_iter():
-        if process.username() == user:
-            result.append(process.pid)
-    return result
+    try:
+        for process in psutil.process_iter(["name", "exe", "cmdline"]):
+            if process.username() == user:
+                yield process
+    except (psutil.AccessDenied, psutil.NoSuchProcess):
+        pass
+
 
+def kill_user_processes(user, sig=signal.SIGTERM, timeout=5, sure_kill=False):
+    """Kill all processes for a specific user.
 
-def kill_user_processes(user):
-    """Kill all processes for a specific user
+    Args:
+        user (str): The user name to match against.
+        sig (signal, optional): See signal module constants. Defaults to signal.SIGTERM.
+        timeout (int, optional): How long to wait for a process to terminate (seconds) before raise exception or, if sure_kill=True, send a SIGKILL. Defaults to 5.
+        sure_kill (bool, optional): Whether to fallback to SIGKILL if the timeout exceeded for the terminate operation or not. Defaults to False.
 
-    Arguments:
-        user {str} -- username
+    Returns:
+        list of psutil.Process or None): list of process objects that remain alive if any. None otherwise.
     """
-    for pid in get_user_processes(user):
-        kill(pid)
+    failed_processes = []
+    for proc in get_user_processes(user):
+        try:
+            kill(proc, sig=sig, timeout=timeout, sure_kill=sure_kill)
+        except (j.exceptions.Runtime, j.exceptions.Permission) as e:
+            j.logger.exception("ignoring an exception that occurred while iterating over user processes", exception=e)
+            failed_processes.append(proc)
 
+    # making sure
+    if failed_processes:
+        gone, failed_processes = psutil.wait_procs(failed_processes, timeout=0)
 
-def get_similar_processes():
-    """Gets similar processes to current process
+    j.logger.debug(f"stay alive: {len(failed_processes)}, pids -> {[p.pid for p in failed_processes]}")
+    return failed_processes or None  # return None if the failed list is empty
```

Comment:
No need to return None, just return empty list.

Reply:
the same as the previous replay on similar comments.